src/sentry/tasks/split_discover_dataset.py

```diff
@@ -0,0 +1,226 @@
+    DashboardWidget,
+    DashboardWidgetQuery,
+    DashboardWidgetTypes,
+)
+from sentry.search.events.builder import QueryBuilder
+from sentry.search.events.types import ParamsType, QueryBuilderConfig
+from sentry.snuba.dataset import Dataset
+from sentry.snuba.referrer import Referrer
+from sentry.tasks.base import instrumented_task
+from sentry.utils import metrics
+from sentry.utils.cache import cache
+from sentry.utils.query import RangeQuerySetWrapper
+
+logger = logging.getLogger("sentry.tasks.split_discover_dataset")
+
+TASK_QUERY_PERIOD = "30m"
+DASHBOARD_QUERY_PERIOD = "1h"
+
+
+def _get_widget_processing_batch_key() -> str:
+    return "on-demand-metrics:widgets:currently-processing-batch"
+
+
+def _set_currently_processing_batch(current_batch: int) -> None:
+    cache.set(_get_widget_processing_batch_key(), current_batch, timeout=3600)
+
+
+def _get_previous_processing_batch() -> int:
+    return cache.get(_get_widget_processing_batch_key(), 0)
+
+
+def _get_current_processing_batch(total_batches: int) -> int:
+    previous_batch = _get_previous_processing_batch()
+    current_batch = (previous_batch + 1) % total_batches
+    return current_batch
+
+
+def _get_batch_for_widget(widget_query_id: int, total_batches: int) -> int:
+    return widget_query_id % total_batches
+
+
+class SplitDiscoverDatasetException(Exception):
+    pass
+
+
+@instrumented_task(
+    name="sentry.tasks.split_discover_dataset",
+    queue="split_discover_dataset",
+    max_retries=0,
+    soft_time_limit=150,
+    time_limit=180,
+    expires=360,
+)
+def schedule_widget_discover_split():
+    """
+    This task is a part of an effort to split the current "Errors and Transactions" widget
+    that queries Discover, into separate widgets for errors and transactions. This task checks
+    if the widgets current data is sourced from errors, transactions or both, and updates the
+    split_discover_dataset field accordingly.
+
+    It does not perform the actual migration of data, but only schedules the update of the
+    split_discover_dataset field which will be used to perform the actual migration from
+    discover dataset to error or transaction dataset in the future.
+
+    # Ops
+    Killswitch option: `split_discover_dataset.enable`
+    """
+    if not options.get("split_discover_dataset.enable"):
+        return
+
+    logger.log(logging.INFO, "Scheduling widget discover split task")
+
+    rollout = options.get("split_discover_dataset.rollout")
+    total_batches = options.get("split_discover_dataset.query.total_batches")
+    widgets_per_batch = options.get("split_discover_dataset.query.batch_size")
+
+    currently_processing_batch = _get_current_processing_batch(total_batches)
+
+    widget_ids = []
+    dashboard_widget_count = 0
+
+    # collect widgets that need to be updated
+    for (widget_id,) in RangeQuerySetWrapper(
+        DashboardWidget.objects.filter(
+            widget_type=DashboardWidgetTypes.DISCOVER, discover_widget_split=None
```

Comment:
Might as well add another option here to exclude the `discover_widget_split=None` since it's possible we may need to override these if we make any mistakes, overriding should be fine up until we use this column to migrate.